연락처 관리 앱 [STEP 1] NERO, Nnn

Nnn-Park opened the pull request:
# 연락처 관리 앱 프로젝트
안녕하세요, 그린(@GREENOVER)!

STEP 1 PR 보내 드립니다! 
잘 부탁드립니다. 🫡

👥 네로(@ODDNERO), 앤(@Nnn-Park) 

<br />

## 🗂️ 폴더 구조

```
ios-contact-manager-ui/
├── 📂 Controller/
│ └── 📄 ViewController.swift
├── 📂 Model/
│ ├── 📄 ContactManager.swift
│ └── 📄 Person.swift
└── 📂 View/
  ├── 📄 LaunchScreen.storyboard
  └── 📄 Main.storyboard
```  
<br />

## 🤖 구성 (MVC)

### 모델
- 싱글톤 패턴으로 데이터 생성
- 데이터 타입: `Person` / 프로퍼티: `name`, `age`, `phoneNumber` 
- `ContactManager`를 통해 CRUD 메서드 구현
Create: `addPerson(person:)`
Read: `readContacts()`
Update: `editContact(inputName:inputAge:inputPhone:)`
Delete: `deletePerson(inputName:)`

### 뷰
- Storyboard에 Table View 생성
- Table View에 Subtitle Cell 생성

### 컨트롤러
- `UITableViewDataSource`를 채택하여 Table View, Cell을 관리
- `.dequeueReusableCell`을 통해 Cell을 재사용

<br />


## 📝 참고 사항

모델 기능 부분은 기본 로직만 구현된 상태
-> 다음 Step부터 UUID를 적용해 볼 예정입니다!

 <br />


## ❓ 질문 사항

1. 내부 저장 방식을 JSON으로 하면 좋을지, CoreData로 하면 좋을지 고민 중에 있습니다. 현업에서는 주로 어떤 방식을 사용하는지 궁금합니다! 
2. 현재 Storyboard로 뷰를 구현한 상태인데, 현업에서도 Storyboard를 많이 사용하는지 궁금합니다. 🧐
3. 이외에 리팩토링이 필요한 부분들이나, 프로젝트를 진행함에 있어 참고하면 좋을 사항들이 있다면 마구마구 말씀해 주세요! 👀👂🏻

 <br />



## 📸 스크린샷
<img width="300" alt="capture01" src= https://github.com/tasty-code/ios-contact-manager/assets/75471842/d651346d-781b-43ee-b890-6cc7eb1f2793>

ODDNERO committed:
config: 프로젝트 생성

ODDNERO committed:
config: gitignore 파일 생성

gitignore에 .DS_Store 파일 추가

ODDNERO committed:
feat: 스토리보드에 테이블뷰 추가

테이블뷰 레이아웃 설정

뷰컨트롤러에서 UITableViewDataSource 채택

ODDNERO committed:
feat. tableViewCell 추가

뷰컨트롤러를 확장하여 UITableViewDataSource 채택한 후 cell 정의

ODDNERO committed:
feat. Person 데이터 타입, 연락처 관리 모델 ContactManager 추가

모델 기능 로직 기본 구현

ODDNERO committed:
chore: 폴더링

MVC 패턴 적용

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/Person.swift:
```diff
@@ -0,0 +1,14 @@
+
+import UIKit
```
UIKit을 임포트하신 이유가 있으실까요~?
UI 컴포넌트를 사용하지 않는 모델 타입으로 보여서요🤔

ODDNERO replied:
단순하게 UIKit로 모두 통일하여 임포트했었습니다 😅😅
항상 용도에 맞게 지정하는 것이 좋을까요? 만약 그렇다면 어떤 점에서 좋은 것인지 궁금합니다!
성능상 차이가 있는지? 더 적재적소에 쓰여 적절하기 때문인지? 🤔

GREENOVER replied:
성능보다는 채택하는 역할 초점에 맞춰보면 좋을것 같아요 🙋🏻
UIKit은 이미 Foundation을 채택하고 있기에 UIKit을 채택했을때는 뷰나 인터페이스를 구현한다거나 하는것들이 사용하지 않아도 추가되있기에 말씀드렸습니다~!
Foundation 프레임워크는 데이터 관리나 네트워크, 파일 액세스 등의 기능을 가진 프레임워크로 사실상 현재 코드에서는 Foundation 채택도 필요하진 않긴합니다!
다만 해당 파일에서 추후 데이터 관리나, 네트워크 , 파일 액세스 등 Foundation의 기능을 사용할 여지가 있을 수 있어 Foundation을 import해놓는것도 적절할것 같지만, UIKit을 import했을때 활용하지 않기에 그 부분은 조금 차이가 있을것 같아요..!

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/ContactManager.swift:
```diff
@@ -0,0 +1,53 @@
+
+import UIKit
```
Person 파일과 동일하게 UIKit을 임포트하신 이유 궁금합니다~!

ODDNERO replied:
위와 동일합니다!

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/ContactManager.swift:
```diff
@@ -0,0 +1,53 @@
+import UIKit
+
+class ContactManager {
+    static let shared = ContactManager()
```
싱글턴의 단점은 뭘까요? 🙋🏻
의존성 주입이 아닌 싱글턴으로 하게 될때 여러분들이 생각하시는 장단점과 채택 이유가 궁금합니다ㅎㅎ

ODDNERO replied:
- **Singleton 장단점**
한 번의 생성으로 여러 곳에서 사용할 수 있어 편리하다는 장점이 있습니다
그러나 코드 간에 결합도와 의존성이 높아져 유지 보수에 어려움이 생기게 되고, 기능을 확장할 때마다 많은 수정이 필요해질 수 있습니다

- **채택 이유**
아직 내부 저장 데이터를 구현하기 전이어서 편의상 임시적으로 Singleton 패턴을 적용했습니다!

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/ContactManager.swift:
```diff
@@ -0,0 +1,53 @@
+
+class ContactManager {
+    static let shared = ContactManager()
+        private init() {}
```
인덴트가 4칸 더 들어간것 같아요..!
다른 코드들과 코드 컨벤션을 맞추면 어떨까요~?

ODDNERO replied:
미처 못 찾았던 부분...! 짚어 주셔서 감사합니다 🙇🏻‍♀️

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/ContactManager.swift:
```diff
@@ -0,0 +1,53 @@
+    static let shared = ContactManager()
+        private init() {}
+    
+    private var contactList: [Person] = [
```
해당 값들은 단순 테스트용 데이터이기에 별도로 구분해보면 어떨까 싶습니다..!
```swift
extension ContactManager {
  static let stub: [Person] = [...]
}
```
요런식도 있을것 같고 다른 방식도 많을것 같아요ㅎㅎ
추후 실제 데이터를 사용해야 한다면 현재 코드들을 변경해아하기에
stub 데이터를 별도로 만들고 추후 VC에서 갈아끼우기만하면 유지보수가 더 쉬울것 같습니다 🙋🏻

ODDNERO replied:
stub이라고 따로 칭하면 좋겠군요! 이 부분도 적용해 보겠습니다 🫡

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/ContactManager.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+    
+    func deletePerson(inputName: String) {
+        if let index = contactList.firstIndex(where: { $0.name == inputName }) {
```
만약 name이 같은 사용자가 여럿 있다면 현재 name으로 같은 사용자 정보인지 체크하는것이 괜찮을까요..?
Person 모델 객체에 UUID 값을 부여하는건 어떤지 제안드립니다 🙋🏻

ODDNERO replied:
Step 2에서 UUID를 적용해 보겠습니다! 🫡🫡

GREENOVER commented on .gitignore:
```diff
@@ -0,0 +1,179 @@
+# Created by https://www.toptal.com/developers/gitignore/api/macos,swift,objective-c,xcode,swiftpackagemanager,cocoapods
```
실제 현재 프로젝트에서 gitignore를 통해 얻고자하는것은 어떤것이셨는지 궁금해요! 🙋🏻
해당 프로젝트에서 어떤 부분들이 실제 gitignore되고 있으신가요~?

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ViewController.swift:
```diff
@@ -0,0 +1,34 @@
+
+import UIKit
+
+class ViewController: UIViewController {
```
조금 더 명확한 뷰컨트롤러 네이밍은 어떨까요~?
추후 만약 더 많은 VC가 생기게되면 네이밍 수정이 일어나야할것 같아 보여서요ㅎㅎ
해당 VC가 어떤 VC인지 조금 더 명확하면 좋을것 같습니다!
또한 final 키워드도 한번 학습해보시면 좋겠습니다 🙋🏻

ODDNERO replied:
앞으로는 뷰컨트롤러 이름을 명확하게 지어 보겠습니다!
ContactListViewController로 네이밍 수정하였고, final 키워드는 아래 답변에 같이 달겠습니다 🙋🏻

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ViewController.swift:
```diff
@@ -0,0 +1,34 @@
+import UIKit
+
+class ViewController: UIViewController {
+    let model = ContactManager.shared
```
private해도 되지 않을까요? 🤔

ODDNERO replied:
해당 부분 이렇게 수정하겠습니다!
```swift
final class ContactListViewController: UIViewController {
    private let model = ContactManager.shared
```

### ❓**궁금한 점**❓
final과 private이 어떤 것인지는 공부했는데, 두 키워드를 쓸 때 적절한 기준점을 잘 모르겠습니다 🤔
1. `final`: 이후에 상속하지 않을 것 같은 class면 일단 붙이면 되는 걸까요?
2. `private`: 다른 파일에서 쓰지 않을 것 같거나, 외부에서 바로 접근하지 않을 것 같은 멤버면 다 붙이고 이후 제약이 생겼을 때 떼면 될지?... 스스로 기준을 잡기가 아직 어려워서 애매하네요 🥲

GREENOVER replied:
final을 사용하면 성능 개선에 도움을 줄 수 있는지도 확인해보면 좋습니다..!
private 사용도 당연히 현재 사용되진 않는다고 확신되지만 추후 변경 지점이 생겨 사용될 경우가 있다면 말씀하신것처럼 추후 변경해도 괜찮아요ㅎㅎ
두 키워드에서 고민이라면 접근보다는 상속에 포인트를 맞춰보고 두개를 고려해봐도 좋을것 같아요!

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ViewController.swift:
```diff
@@ -0,0 +1,34 @@
+    @IBOutlet weak var tableView: UITableView!
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
```
실제 해당 코드는 어떤 일을 해주나요~?
현재 코드에서는 없어도 정상적으로 동작하고 보여지는데 왜 필요할까요!?

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ViewController.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+}
+    
+extension ViewController: UITableViewDataSource {
```
extension으로 분기한것 아주 좋습니다!
가독성이 너무 좋아요ㅎㅎ 👍

ODDNERO replied:
앞으로도 extension을 적극 활용해야겠어요! 감사합니다 😸

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Controller/ViewController.swift:
```diff
@@ -0,0 +1,34 @@
+        cell.textLabel?.text = "\(personIndex.name)(\(personIndex.age))"
+        cell.detailTextLabel?.text = "\(personIndex.phoneNumber)"
+        cell.accessoryType = .disclosureIndicator
+        
```
한칸 개행해주신 이유가 단순히 궁금합니다~!
cell을 set해주는 느낌들의 위 코드들은 개행이 없는데 selectionStyle 속성만 부여할때 개행이 되어 있어서 궁금했어요ㅎㅎ

ODDNERO replied:
같은 cell 관련 코드지만, 네 줄을 연속으로 두는 것보다는 각 기능에 따라 구분해 주고 싶었습니다!
대부분이 새로 공부하게 된 것들이라, 세세하게 나눠 놓지 않으면 뭐가 뭐였는지 아직 헷갈리더라구요 😂

1. cell 위에 새로운 정보를 보여 주는 코드
2. cell의 단순 속성(컬러, 스타일)을 담고 있는 코드
❓ `.disclosureIndicator`를 액세서리 뷰의 속성으로 볼 수도 있지만, `>` 모양의 버튼을 추가로 보여 준다는 느낌이 있어 1로 분류했습니다

GREENOVER commented on ios-contact-manager-ui/ios-contact-manager-ui/Model/ContactManager.swift:
```diff
@@ -0,0 +1,53 @@
+
+import UIKit
+
+class ContactManager {
```
모델 그룹에 들어가도 괜찮긴 하지만 뭔가 하는 역할이 모델들을 다루는 그런 역할로 보이는데요..!
답은 없지만, 저의 경우는 해당 Manager와 같은 역할을 하는 객체들은 Service 그룹으로 묶어 관리하긴 합니다ㅎㅎ
모델에는 딱 모델 타입들만 들어갈 수 있도록 관심사를 나누고 있어요.
다시 말씀드리지만, 하나의 시야일뿐..! 잘못된것이 아니니 이런것도 있구나~ 하고 넘어가셔도 좋습니다..!!

ODDNERO replied:
오 Service라는 그룹으로 묶으면 더 적절하군요...!!
폴더링을 제 힘으로 처음 해 봐서 딱 모델/뷰/컨트롤러만 생각한 상태로 나눠 봤었습니다 ㅎㅎㅎ
모델 관리 하는 역할을 모델과 따로 나눌 수 있다는 점 새로 알고 갑니다! 👍🏻

GREENOVER reviewed (changes requested):
@Nnn-Park @ODDNERO 
두분 스텝 1 너무 수고 많으셨습니다!!
Step1의 요구사항은 충분히 잘 담고 계시고 구현도 잘하셔서 추가로 보완되면 좋을 부분들과 정말 궁금한 부분들에 대해 주로 리뷰 코멘트 남겼습니다ㅎㅎ
***
### 칭찬 드리고 싶은 부분
- 전체적인 코드 가독성이 좋아요 💯
- 파일, 그룹 및 코드를 관심사에 맞게 잘 나눠주셔서 보기 수월합니다 💯
***
### 궁금하신 부분에 대한 저의 의견
- 내부 저장 방식을 JSON으로 하면 좋을지, CoreData로 하면 좋을지 고민 중에 있습니다. 현업에서는 주로 어떤 방식을 사용하는지 궁금합니다!
  - 물론 프로젝트나 팀마다 그리고 상황마다 다 다르겠지만, 저는 주로 내부에 저장해야한다면, UserDefault와 CoreData 그리고 Realm을 주로 사용합니다.
  여기서 Realm은 CoreData와 같이 내부 저장소에 데이터를 저장하도록 해주는 외부 라이브러리이긴하지만요..!
  JSON으로는 사용할 수도 있겠지만, 흔히 그렇게는 잘 사용하지 않기에 CoreData쪽으로 학습해보셔도 좋을것 같아요ㅎㅎ
- 현재 Storyboard로 뷰를 구현한 상태인데, 현업에서도 Storyboard를 많이 사용하는지 궁금합니다. 🧐
  - 이것도 정말 팀과 프로젝트마다 다르겠지만, 주로 많은곳에서 스토리보드를 지양하고 코드베이스로 작업하려 하긴 합니다!
  스토리보드로 UI를 그리는것과 코드로 UI를 그리는것에 대해 장단점들은 아마 아실것 같긴해요!
  호옥시 아직 학습전이시면 학습해보셔도 좋습니다.
  결론적으로 저의 생각은 여러명의 협업시에는 코드 베이스로 작업하는것이 더 적절할것 같아요!
- 이외에 리팩토링이 필요한 부분들이나, 프로젝트를 진행함에 있어 참고하면 좋을 사항들이 있다면 마구마구 말씀해 주세요! 👀👂🏻
  - 요 부분들에 대해 리뷰 코멘트로 남겨봤습니다!ㅎㅎ
  이미 잘 나눠주시고 구현해주셔서 현재 단계에서 크게는 없지만요 😃
***
두분 다시 한번 고생 많으셨어요ㅎㅎ
날씨가 추워지고 폭설인데... 항상 건강 관리 최우선으로 하시길 바래요~!
우선은 스텝 1 요구사항은 다 만족하고 동작에도 이상이 없으나 한번 점검차 Request Changes로 드리겠습니다..!
언제든 머지 혹은 재리뷰 요청을 해주세요 🙇🏻

ODDNERO committed:
design: DetailViewController 생성 및 스토리보드에 Detail View 추가

- NavigationBar 추가
- NavigationItem 추가
- DetailViewController 연결
- TextField AutoLayout 설정

ODDNERO committed:
chore: 코드 정렬 및 ViewController 네이밍 변경

ViewController -> ContactListViewController

ODDNERO committed:
feat: 버튼을 통한 모달 연결 구현 및 키보드 타입 지정

UITextFieldDelegate 채택

ODDNERO committed:
chore: 코드 정렬
